src/devtools/client/inspector/rules/components/RulesApp.tsx

```diff
@@ -63,120 +70,123 @@ export const RulesApp: FC<RulesAppProps> = ({
     showContextMenu(event);
   };
 
-  const renderInheritedRules = (rules: RuleState[]) => {
-    if (!rules.length) {
-      return null;
-    }
+  const renderInheritedRules = useCallback(
+    (rules: InheritedRule[]) => {
+      const output = [];
+      let lastInheritedNodeId;
 
-    const output = [];
-    let lastInheritedNodeId;
+      for (const rule of rules) {
+        const { inheritedNodeId, inheritedSource } = rule.inheritance;
 
-    for (const rule of rules) {
-      if (!rule.inheritance) {
-        continue;
-      }
+        if (inheritedNodeId !== lastInheritedNodeId) {
+          lastInheritedNodeId = inheritedNodeId;
+          output.push(
+            <div key={inheritedNodeId} className="ruleview-header">
+              {inheritedSource}
+            </div>
+          );
+        }
 
-      const { inheritedNodeId, inheritedSource } = rule.inheritance;
-
-      if (inheritedNodeId !== lastInheritedNodeId) {
-        lastInheritedNodeId = inheritedNodeId;
         output.push(
-          <div key={inheritedNodeId} className="ruleview-header">
-            {inheritedSource}
-          </div>
+          <Rule
+            key={`${inheritedNodeId}|${rule.id}`}
+            rule={rule}
+            {...ruleProps}
+            query={rulesQuery}
+          />
         );
       }
 
-      output.push(
-        <Rule key={`${inheritedNodeId}|${rule.id}`} rule={rule} {...ruleProps} query={rulesQuery} />
-      );
-    }
-
-    return output;
-  };
-
-  const renderStyleRules = (rules: RuleState[]) => {
-    if (!rules.length) {
-      return null;
-    }
-
-    return <Rules {...ruleProps} rules={rules} />;
-  };
+      return output;
+    },
+    [rulesQuery, ruleProps]
+  );
 
-  const renderPseudoElementRules = (rules: RuleState[]) => {
-    if (!rules.length) {
-      return null;
-    }
+  const renderStyleRules = useCallback(
+    (rules: RuleState[]) => {
+      return <Rules {...ruleProps} query={rulesQuery} rules={rules} />;
+    },
+    [rulesQuery, ruleProps]
+  );
 
-    type FCProps<C> = C extends FC<infer P> ? P : never;
-    const componentProps: FCProps<typeof Rules> = {
-      rules,
-      ...ruleProps,
-    };
-
-    const items = [
-      {
-        component: Rules,
-        componentProps,
-        header: getStr("rule.pseudoElement"),
-        id: "rules-section-pseudoelement",
-        opened: Services.prefs.getBoolPref(SHOW_PSEUDO_ELEMENTS_PREF),
-        onToggle: (opened: boolean) => {
-          Services.prefs.setBoolPref(SHOW_PSEUDO_ELEMENTS_PREF, opened);
+  const renderPseudoElementRules = useCallback(
+    (rules: RuleState[]) => {
+      type FCProps<C> = C extends FC<infer P> ? P : never;
```

Comment:
That is such a useful type, perhaps we should move it to some utilities file

Reply:
Because our codebase rarely uses `FC` components and we still have a lot of untyped JS, it's unlikely that many of these cases will come up. Eventually we'll migrate everything to TS and we won't have to do inference ourselves but TS typecheker will be able to warn us (i.e. when we migrate the `Accordion` component). I'll refrain from complicating it for now.